src/foam/box/SocketService.js

```diff
@@ -0,0 +1,183 @@
+      name: 'parser',
+      javaType: 'foam.lib.json.JSONParser',
+      javaFactory: `return getX().create(JSONParser.class);`
+    },
+    {
+      class: 'Object',
+      name: 'router',
+      javaType: 'foam.nanos.box.NanoServiceRouter',
+      javaFactory: `
+        return getX().create(NanoServiceRouter.class);
+      `
+    }
+  ],
+
+  constants: [
+    {
+      type: 'Integer',
+      name: 'BUFFERSIZE',
+      value: 1024
+    },
+  ],
+
+  methods: [
+    {
+      name: 'setupServer',
+      args: [
+        {
+          name: 'port',
+          type: 'Integer'
+        }
+      ],
+      javaCode:
+        `try {
+          ServerSocket serverSocket = new ServerSocket(port);
+          setServerSocket(serverSocket); 
+          setPort_(port);
+          while (true) {
+            new SocketClientHandler(getX(), getServerSocket().accept(), getParser(), getRouter()).start();
+          }
+        } catch (IOException e) {
+          port = 10000 + (new Random().nextInt(10000) + 1);
+          setupServer(port);
+        }`
+    },
+    {
+      name: 'stop',
+      javaCode:
+        `try {
+          getServerSocket().close();
+        } catch (IOException ex) {
+          ((Logger) getX().get("logger")).error(ex);
+        }`     
+    }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(foam.java.Code.create({
+          data:
+            `
+            private static class SocketClientHandler extends Thread {
+              private X x;
+              private Socket clientSocket;
+              private JSONParser parser;
+              private NanoServiceRouter router;
+          
+              public SocketClientHandler(X x, Socket socket, JSONParser parser, NanoServiceRouter router) {
+                this.x = x;
+                this.parser = parser;
+                this.clientSocket = socket;
+                this.router = router;
+              }
+
+              public void run() {
+                try {
+                  String mStr = Base64.getEncoder().encodeToString(read(this.clientSocket));
+                  Message message = (Message) this.parser.parseString(mStr);
+                  this.router.send(message);
+                } catch (java.lang.Exception ex) {
+                  ((Logger) this.x.get("logger")).error(ex);
+                } finally {
+                  try {
+                    clientSocket.close();
+                  } catch (IOException e) {
+                    ((Logger) this.x.get("logger")).error(e);
+                  }
+                }
+              }
+            }
+            public static byte[] read(Socket socket) throws IOException{
+              byte[] buffer = new byte[BUFFERSIZE];
+              ByteArrayOutputStream data = new ByteArrayOutputStream();
+              int read;
+              try {
+                while ((read = socket.getInputStream().read(buffer, 0, buffer.length)) != -1) {
```

Comment:
This current plan only appears to let us send one message per socket.  As we're reading until reading fails and then trying to parse.

What we want instead is to add some sort of framing support to the socket so we can read a possibly inifinite sequence of distinct objects.  In JS we accomplished this here:

https://github.com/foam-framework/foam2/blob/master/src/foam/net/node/Socket.js#L92

Before sending each message over the socket, we send the size of the message to be sent in bytes . Then the receiver here:

https://github.com/foam-framework/foam2/blob/master/src/foam/net/node/Socket.js#L136

buffers data until it's received a full payload, parses and dispatches that payload, and then begins buffering the next message.